custom_components/econnect_metronet/switch.py

```diff
@@ -0,0 +1,95 @@
+"""Module for e-connect switch (outputs)."""
+from elmo import query as q
+from homeassistant.components.switch import SwitchEntity
+from homeassistant.config_entries import ConfigEntry
+from homeassistant.core import HomeAssistant
+from homeassistant.helpers.entity_platform import AddEntitiesCallback
+from homeassistant.helpers.update_coordinator import (
+    CoordinatorEntity,
+    DataUpdateCoordinator,
+)
+
+from .const import (
+    CONF_EXPERIMENTAL,
+    CONF_FORCE_UPDATE,
+    DOMAIN,
+    KEY_COORDINATOR,
+    KEY_DEVICE,
+)
+from .devices import AlarmDevice
+from .helpers import generate_entity_id
+
+
+async def async_setup_entry(
+    hass: HomeAssistant,
+    entry: ConfigEntry,
+    async_add_entities: AddEntitiesCallback,
+) -> None:
+    """Set up e-connect binary sensors from a config entry."""
+    device = hass.data[DOMAIN][entry.entry_id][KEY_DEVICE]
+    coordinator = hass.data[DOMAIN][entry.entry_id][KEY_COORDINATOR]
+    # Load all entities and register outputs
+
+    outputs = []
+
+    # Iterate through the outputs of the provided device and create OutputSwitch objects
+    for output_id, name in device.outputs:
+        unique_id = f"{entry.entry_id}_{DOMAIN}_{q.OUTPUTS}_{output_id}"
+        outputs.append(OutputSwitch(unique_id, output_id, entry, name, coordinator, device))
```

Comment:
Should you do that though? I mean, can't you check if actually you have the required setting turned on, and add the switch (so the entity) only if the setting is there (so `control_denied_to_users` and `do_not_require_authentication`?

In this way, we avoid to add entities you can't use anyway. Maybe you can add a `_LOGGER.info` message to mention why it is not enabled and how they can enable it.

What do you think?

Reply:
For my opinion, it would be better to leave even the non-handled outputs as commands because they can be useful as a status, and it would be misleading to categorize them as binary sensors

Reply:
The issue is that we don't know if the most common configuration is the one that makes it work. If most users can't use the switch, we'll just receive complaints that the functionality doesn't work, as they will encounter switches that they click with no effect. Also, if users check their logs, they will see an error without knowing what to do next, other than opening bug reports for something we can't handle.

To prevent being overwhelmed by bugs we can't fix, I would recommend at least adding a `_LOGGER.info()` to explain that the switch will not work because [insert the reason and what they should do to enable it (or what to ask from the vendor)].

Reply:
I would propose the following revisions:
1. We can maintain the loop and still create the switches, even if they are not functional due to the settings being disabled in the central unit.
2. Instead of displaying an error when the switch is used, show a warning (`_LOGGER.warning()`) that explains why the functionality is disabled and what steps can be taken to enable it (including checking with the vendor who installed the CU, if applicable).
3. Since people do not see logs when they activate the switch, they might experience it turning on and off after a brief delay (if it doesn't work, it will revert to the original state). This could lead to multiple attempts before realizing it doesn't work, which is not ideal for the user experience.
4. We should explore the possibility of using Home Assistant popups to notify users in the UI directly. Maybe you can check other integrations or the documentation to see how you can raise a message to show the actual popup.

What do you think?

Reply:
Your idea is great, but I don't know exactly how to display the popup notification. 
Have you already found something in other integrations?

Reply:
Let me take a gander, I'll check it tomorrow!

Reply:
It looks like we don't have any other official way to send a notification other than using the `persistent_notification` service. So let's use it:
1. `Device.turn_on` should return a boolean, meaning `True` if it goes well or `False` if it doesn't
2. In the switch (`async_turn_on`) you should update the code so that you check the result and send a notification if the change didn't go through:
```python
from homeassistant.components import persistent_notification
# [...]

async def async_turn_on(self):
    """Turn the entity on."""
    if not await self.hass.async_add_executor_job(self._device.turn_on, self._output_id):
        persistent_notification.async_create(self.hass, message, title, identifier)
```

The parameters meaning is:
- `message`: the body message. You should put there the details about "why" it's not working, and what users can do to enable it (or what they should ask the vendor). This information is crucial.
- `title`: the title of the notification. It can be something along the lines "Unable to toggle the switch". Up to you.
- `identifier`: this is a static value that is used to avoid creating hundreds of notifications. You can use something like `econnect_metronet_output_fail`.

I would recommend to not put these values in this code-block, but to put them in the `const.py` so that you can reuse them in the code. Maybe, you can use the same `message` in the `_LOGGER.warning()`, so that you explain what's going on even in logs.

For now, write a draft message in English, we'll make another PR to figure out how to use translations.